File: ext/scheduler.cpp
#include <iostream>
#include "scheduler.h"

using namespace mesos;
namespace HsFFI {

class HsScheduler : public Scheduler
{
public:
	HsScheduler(
		OnSchedulerRegisteredCallback* registeredCb,
		OnSchedulerReRegisteredCallback* reRegisteredCb,
		OnSchedulerDisconnectedCallback* disconnectedCb,
		OnSchedulerResourceOffers* resourceOffersCb,
		OnOfferRescinded* offerRescindedCb,
		OnStatusUpdate* statusUpdateCb,
		OnFrameworkMessage* frameworkMessageCb,
		OnSlaveLost* slaveLostCb,
		OnExecutorLost* executorLostCb,
		OnSchedulerError* schedulerErrorCb)
	{
		this->registeredCb = registeredCb;
		this->reRegisteredCb = reRegisteredCb;
		this->disconnectedCb = disconnectedCb;
		this->resourceOffersCb = resourceOffersCb;
		this->offerRescindedCb = offerRescindedCb;
		this->statusUpdateCb = statusUpdateCb;
		this->frameworkMessageCb = frameworkMessageCb;
		this->slaveLostCb = slaveLostCb;
		this->executorLostCb = executorLostCb;
		this->schedulerErrorCb = schedulerErrorCb;
	}

	~HsScheduler(){}

	virtual void registered (SchedulerDriver* driver,
			const FrameworkID& frameworkId,
			const MasterInfo& masterInfo)
	{
		registeredCb(driver, &frameworkId, &masterInfo);
	}

	virtual void reregistered (SchedulerDriver* driver,
			const MasterInfo& masterInfo)
	{
		reRegisteredCb(driver, &masterInfo);
	}

	virtual void disconnected (SchedulerDriver* driver)
	{
		disconnectedCb(driver);
	}

	virtual void resourceOffers (SchedulerDriver* driver,
			const std::vector<Offer>& offers)
	{
		std::vector<Offer*> pointerized = std::vector<Offer*>(offers.size());
		for (int i = 0; i < offers.size(); ++i)
			pointerized[i] = (Offer*) &offers[i];

		resourceOffersCb(driver, pointerized.data(), pointerized.size());
	}

	virtual void offerRescinded (SchedulerDriver* driver,
			const OfferID& offerId)
	{
		offerRescindedCb(driver, &offerId);
	}

	virtual void statusUpdate (SchedulerDriver* driver,
			const TaskStatus& status)
	{
		statusUpdateCb(driver, &status);
	}

	virtual void frameworkMessage (SchedulerDriver* driver,
			const ExecutorID& executorId,
			const SlaveID& slaveId,
			const std::string& data)
	{
		frameworkMessageCb(driver, &executorId, &slaveId, data.data(), data.size());
	}

	virtual void slaveLost (SchedulerDriver* driver,
			const SlaveID& slaveId)
	{
		slaveLostCb(driver, &slaveId);
	}

	virtual void executorLost (SchedulerDriver* driver,
			const ExecutorID& executorId,
			const SlaveID& slaveId,
			int status)
	{
		executorLostCb(driver, &executorId, &slaveId, status);
	}

	virtual void error (SchedulerDriver* driver, const std::string& message)
	{
		schedulerErrorCb(driver, message.data(), message.size());
	}

private:
	OnSchedulerRegisteredCallback* registeredCb;
	OnSchedulerReRegisteredCallback* reRegisteredCb;
	OnSchedulerDisconnectedCallback* disconnectedCb;
	OnSchedulerResourceOffers* resourceOffersCb;
	OnOfferRescinded* offerRescindedCb;
	OnStatusUpdate* statusUpdateCb;
	OnFrameworkMessage* frameworkMessageCb;
	OnSlaveLost* slaveLostCb;
	OnExecutorLost* executorLostCb;
	OnSchedulerError* schedulerErrorCb;
};
}

SchedulerPtr createScheduler(OnSchedulerRegisteredCallback* registeredCb,
	OnSchedulerReRegisteredCallback* reRegisteredCb,
	OnSchedulerDisconnectedCallback* disconnectedCb,
	OnSchedulerResourceOffers* resourceOffersCb,
	OnOfferRescinded* offerRescindedCb,
	OnStatusUpdate* statusUpdateCb,
	OnFrameworkMessage* frameworkMessageCb,
	OnSlaveLost* slaveLostCb,
	OnExecutorLost* executorLostCb,
	OnSchedulerError* schedulerErrorCb)
{
	return new HsFFI::HsScheduler(registeredCb,
			reRegisteredCb,
			disconnectedCb,
			resourceOffersCb,
			offerRescindedCb,
			statusUpdateCb,
			frameworkMessageCb,
			slaveLostCb,
			executorLostCb,
			schedulerErrorCb
	);
}

void destroyScheduler(SchedulerPtr scheduler)
{
	delete scheduler;
}

void exerciseMethods(SchedulerPtr scheduler)
{	
	scheduler->registered(NULL, FrameworkID(), MasterInfo());
	scheduler->reregistered(NULL, MasterInfo());
	scheduler->disconnected(NULL);
	scheduler->resourceOffers(NULL, std::vector<Offer>());
	scheduler->offerRescinded(NULL, OfferID());
	scheduler->statusUpdate(NULL, TaskStatus());
	scheduler->frameworkMessage(NULL, ExecutorID(), SlaveID(), std::string());
	scheduler->slaveLost(NULL, SlaveID());
	scheduler->executorLost(NULL, ExecutorID(), SlaveID(), 0);
	scheduler->error(NULL, std::string());
}

SchedulerDriverPtr createSchedulerDriver(SchedulerPtr scheduler, FrameworkInfoPtr framework, char* master, int masterLength)
{
	return new MesosSchedulerDriver(scheduler, *framework, std::string(master, masterLength));
}

SchedulerDriverPtr createSchedulerDriverWithCredentials(SchedulerPtr scheduler, FrameworkInfoPtr framework, char* master, int masterLength, CredentialPtr credential)
{
	return new MesosSchedulerDriver(scheduler, *framework, std::string(master, masterLength), *credential);
}

void destroySchedulerDriver(SchedulerDriverPtr schedulerDriver)
{
	delete schedulerDriver;
}

int startSchedulerDriver(SchedulerDriverPtr schedulerDriver)
{
	return schedulerDriver->start();
}

int stopSchedulerDriver(SchedulerDriverPtr schedulerDriver, bool failover)
{
	return schedulerDriver->stop(failover);	
}

int abortSchedulerDriver(SchedulerDriverPtr schedulerDriver)
{
	return schedulerDriver->abort();
}

int joinSchedulerDriver(SchedulerDriverPtr schedulerDriver)
{
	return schedulerDriver->join();
}

int runSchedulerDriver(SchedulerDriverPtr schedulerDriver)
{
	return schedulerDriver->run();
}

int requestResources(SchedulerDriverPtr schedulerDriver, RequestPtr* request, int requestCount)
{
	std::vector<Request> requests = std::vector<Request>(requestCount);
	for(int i = 0; i < requestCount; ++i)
		requests[i] = *request[i];

	return schedulerDriver->requestResources(requests);
}

int launchTasks(SchedulerDriverPtr schedulerDriver, OfferIDPtr* offer, int offerCount, TaskInfoPtr* taskInfo, int infoCount, FiltersPtr filters)
{
	std::vector<OfferID> offers = std::vector<OfferID>(offerCount);
	for(int i = 0; i < offerCount; ++i)
		offers[i] = *offer[i];

	std::vector<TaskInfo> tasks = std::vector<TaskInfo>(infoCount);
	for(int i = 0; i < infoCount; ++i)
		tasks[i] = *taskInfo[i];

	return schedulerDriver->launchTasks(offers, tasks, *filters);
}

int killTask(SchedulerDriverPtr schedulerDriver, TaskIDPtr task)
{
	return schedulerDriver->killTask(*task);
}

int declineOffer(SchedulerDriverPtr schedulerDriver, OfferIDPtr offerId, FiltersPtr filters)
{
	return schedulerDriver->declineOffer(*offerId, *filters);
}

int reviveOffers(SchedulerDriverPtr schedulerDriver)
{
	return schedulerDriver->reviveOffers();
}

int schedulerDriverSendFrameworkMessage(SchedulerDriverPtr schedulerDriver, ExecutorIDPtr executor, SlaveIDPtr slave, char* msg, int msgLength)
{
	return schedulerDriver->sendFrameworkMessage(*executor, *slave, std::string(msg, msgLength));
}

int reconcileTasks(SchedulerDriverPtr schedulerDriver, TaskStatusPtr* statuses, int statusCount)
{
	std::vector<TaskStatus> tasks = std::vector<TaskStatus>(statusCount);
	for(int i = 0; i < statusCount; ++i)
		tasks[i] = *statuses[i];

	return schedulerDriver->reconcileTasks(tasks);
}

File: src/System/Mesos/Raw/Scheduler.hs
module System.Mesos.Raw.Scheduler where
import           System.Mesos.Internal
import           System.Mesos.Raw.Credential
import           System.Mesos.Raw.ExecutorId
import           System.Mesos.Raw.Filters
import           System.Mesos.Raw.FrameworkId
import           System.Mesos.Raw.FrameworkInfo
import           System.Mesos.Raw.MasterInfo
import           System.Mesos.Raw.Offer
import           System.Mesos.Raw.OfferId
import           System.Mesos.Raw.Request
import           System.Mesos.Raw.SlaveId
import           System.Mesos.Raw.TaskId
import           System.Mesos.Raw.TaskInfo
import           System.Mesos.Raw.TaskStatus
type SchedulerPtr = Ptr Scheduler

data Scheduler = Scheduler
  { schedulerImpl                :: SchedulerPtr
  , rawSchedulerRegistered       :: FunPtr RawSchedulerRegistered
  , rawSchedulerReRegistered     :: FunPtr RawSchedulerReRegistered
  , rawSchedulerDisconnected     :: FunPtr RawSchedulerDisconnected
  , rawSchedulerResourceOffers   :: FunPtr RawSchedulerResourceOffers
  , rawSchedulerOfferRescinded   :: FunPtr RawSchedulerOfferRescinded
  , rawSchedulerStatusUpdate     :: FunPtr RawSchedulerStatusUpdate
  , rawSchedulerFrameworkMessage :: FunPtr RawSchedulerFrameworkMessage
  , rawSchedulerSlaveLost        :: FunPtr RawSchedulerSlaveLost
  , rawSchedulerExecutorLost     :: FunPtr RawSchedulerExecutorLost
  , rawSchedulerError            :: FunPtr RawSchedulerError
  }

-- | Type representing the connection from a scheduler to Mesos. This
-- handle is used both to manage the scheduler's lifecycle (start
-- it, stop it, or wait for it to finish) and to interact with Mesos
-- (e.g., launch tasks, kill tasks, etc.).
newtype SchedulerDriver = SchedulerDriver { fromSchedulerDriver :: SchedulerDriverPtr }
  deriving (Show, Eq)

type SchedulerDriverPtr = Ptr SchedulerDriver

type RawSchedulerRegistered = SchedulerDriverPtr -> FrameworkIDPtr -> MasterInfoPtr -> IO ()

type RawSchedulerReRegistered = SchedulerDriverPtr -> MasterInfoPtr -> IO ()

type RawSchedulerDisconnected = SchedulerDriverPtr -> IO ()

type RawSchedulerResourceOffers = SchedulerDriverPtr -> Ptr OfferPtr -> CInt -> IO ()

type RawSchedulerOfferRescinded = SchedulerDriverPtr -> OfferIDPtr -> IO ()

type RawSchedulerStatusUpdate = SchedulerDriverPtr -> TaskStatusPtr -> IO ()

type RawSchedulerFrameworkMessage = SchedulerDriverPtr -> ExecutorIDPtr -> SlaveIDPtr -> Ptr CChar -> Int -> IO ()

type RawSchedulerSlaveLost = SchedulerDriverPtr -> SlaveIDPtr -> IO ()

type RawSchedulerExecutorLost = SchedulerDriverPtr -> ExecutorIDPtr -> SlaveIDPtr -> CInt -> IO ()

type RawSchedulerError = SchedulerDriverPtr -> Ptr CChar -> CInt -> IO ()

foreign import ccall "wrapper" wrapSchedulerRegistered
  :: RawSchedulerRegistered
  -> IO (FunPtr RawSchedulerRegistered)

foreign import ccall "wrapper" wrapSchedulerReRegistered
  :: RawSchedulerReRegistered
  -> IO (FunPtr RawSchedulerReRegistered)

foreign import ccall "wrapper" wrapSchedulerDisconnected
  :: RawSchedulerDisconnected
  -> IO (FunPtr RawSchedulerDisconnected)

foreign import ccall "wrapper" wrapSchedulerResourceOffers
  :: RawSchedulerResourceOffers
  -> IO (FunPtr RawSchedulerResourceOffers)

foreign import ccall "wrapper" wrapSchedulerOfferRescinded
  :: RawSchedulerOfferRescinded
  -> IO (FunPtr RawSchedulerOfferRescinded)

foreign import ccall "wrapper" wrapSchedulerStatusUpdate
  :: RawSchedulerStatusUpdate
  -> IO (FunPtr RawSchedulerStatusUpdate)

foreign import ccall "wrapper" wrapSchedulerFrameworkMessage
  :: RawSchedulerFrameworkMessage
  -> IO (FunPtr RawSchedulerFrameworkMessage)

foreign import ccall "wrapper" wrapSchedulerSlaveLost
  :: RawSchedulerSlaveLost
  -> IO (FunPtr RawSchedulerSlaveLost)

foreign import ccall "wrapper" wrapSchedulerExecutorLost
  :: RawSchedulerExecutorLost
  -> IO (FunPtr RawSchedulerExecutorLost)

foreign import ccall "wrapper" wrapSchedulerError
  :: RawSchedulerError
  -> IO (FunPtr RawSchedulerError)

foreign import ccall safe "ext/scheduler.h createScheduler" c_createScheduler
  :: FunPtr RawSchedulerRegistered
  -> FunPtr RawSchedulerReRegistered
  -> FunPtr RawSchedulerDisconnected
  -> FunPtr RawSchedulerResourceOffers
  -> FunPtr RawSchedulerOfferRescinded
  -> FunPtr RawSchedulerStatusUpdate
  -> FunPtr RawSchedulerFrameworkMessage
  -> FunPtr RawSchedulerSlaveLost
  -> FunPtr RawSchedulerExecutorLost
  -> FunPtr RawSchedulerError
  -> IO SchedulerPtr

foreign import ccall safe "ext/scheduler.h destroyScheduler" c_destroyScheduler
  :: SchedulerPtr
  -> IO ()

foreign import ccall safe "ext/scheduler.h createSchedulerDriver" c_createSchedulerDriver
  :: SchedulerPtr
  -> FrameworkInfoPtr
  -> Ptr CChar
  -> CInt
  -> IO SchedulerDriverPtr

foreign import ccall safe "ext/scheduler.h createSchedulerDriverWithCredentials"  c_createSchedulerDriverWithCredentials
  :: SchedulerPtr
  -> FrameworkInfoPtr
  -> Ptr CChar
  -> CInt
  -> CredentialPtr
  -> IO SchedulerDriverPtr

foreign import ccall safe "ext/scheduler.h destroySchedulerDriver" c_destroySchedulerDriver
  :: SchedulerDriverPtr
  -> IO ()

foreign import ccall safe "ext/scheduler.h startSchedulerDriver" c_startSchedulerDriver
  :: SchedulerDriverPtr
  -> IO CInt

foreign import ccall safe "ext/scheduler.h stopSchedulerDriver" c_stopSchedulerDriver
  :: SchedulerDriverPtr
  -> CInt
  -> IO CInt

foreign import ccall safe "ext/scheduler.h abortSchedulerDriver" c_abortSchedulerDriver
  :: SchedulerDriverPtr
  -> IO CInt

foreign import ccall safe "ext/scheduler.h joinSchedulerDriver" c_joinSchedulerDriver
  :: SchedulerDriverPtr
  -> IO CInt

foreign import ccall safe "ext/scheduler.h runSchedulerDriver" c_runSchedulerDriver
  :: SchedulerDriverPtr
  -> IO CInt

foreign import ccall safe "ext/scheduler.h requestResources" c_requestResources
  :: SchedulerDriverPtr
  -> Ptr RequestPtr
  -> CInt
  -> IO CInt

foreign import ccall safe "ext/scheduler.h launchTasks" c_launchTasks
  :: SchedulerDriverPtr
  -> Ptr OfferIDPtr
  -> CInt
  -> Ptr TaskInfoPtr
  -> CInt
  -> FiltersPtr
  -> IO CInt

foreign import ccall safe "ext/scheduler.h killTask" c_killTask
  :: SchedulerDriverPtr
  -> TaskIDPtr
  -> IO CInt

foreign import ccall safe "ext/scheduler.h declineOffer" c_declineOffer
  :: SchedulerDriverPtr
  -> OfferIDPtr
  -> FiltersPtr
  -> IO CInt

foreign import ccall safe "ext/scheduler.h reviveOffers" c_reviveOffers
  :: SchedulerDriverPtr
  -> IO CInt

foreign import ccall safe "ext/scheduler.h schedulerDriverSendFrameworkMessage" c_sendFrameworkMessage
  :: SchedulerDriverPtr
  -> ExecutorIDPtr
  -> SlaveIDPtr
  -> Ptr CChar
  -> CInt
  -> IO CInt

foreign import ccall safe "ext/scheduler.h reconcileTasks" c_reconcileTasks
  :: SchedulerDriverPtr
  -> Ptr TaskStatusPtr
  -> CInt
  -> IO CInt
